src/core/geometry/qgsgeometryutils.h

```diff
@@ -106,11 +106,36 @@ class CORE_EXPORT QgsGeometryUtils
      * \param p2 First segment end point
      * \param q1 Second segment start point
      * \param q2 Second segment end point
-     * \param inter Output parameter, the intersection point
+     * \param intersectionPoint Output parameter, the intersection point
+     * \param isIntersection Output parameter, return true if an intersection is found
      * \param tolerance The tolerance to use
+     * \param acceptImproperIntersection By default, this method returns true only if segments have proper intersection. If set true, returns also true if segments have improper intersection (end of one segment on other segment ; continuous segments).
      * \returns  Whether the segments intersect
+     * * Example:
+     * \code{.py}
+     *   epsilon = 1e-8
+     *   ret = QgsGeometryUtils.segmentIntersection( QgsPoint( 0, 0 ), QgsPoint( 0, 1 ), QgsPoint( 1, 1 ), QgsPoint( 1, 0 ), epsilon )
+     *   ret[0], ret[1].asWkt(), ret[2]
+     *   # Whether the segments intersect, the intersection point, is intersect
+     *   # (False, 'Point (0 0)', False)
+     *   ret = QgsGeometryUtils.segmentIntersection( QgsPoint( 0, 0 ), QgsPoint( 0, 5 ), QgsPoint( 0, 5 ), QgsPoint( 1, 5 ), epsilon )
+     *   ret[0], ret[1].asWkt(), ret[2]
+     *   # (False, 'Point (0 5)', True)
+     *   ret = QgsGeometryUtils.segmentIntersection( QgsPoint( 0, 0 ), QgsPoint( 0, 5 ), QgsPoint( 0, 5 ), QgsPoint( 1, 5 ), epsilon, True )
+     *   ret[0], ret[1].asWkt(), ret[2]
+     *   # (True, 'Point (0 5)', True)
+     *   ret = QgsGeometryUtils.segmentIntersection( QgsPoint( 0, 0 ), QgsPoint( 0, 5 ), QgsPoint( 0, 2 ), QgsPoint( 1, 5 ), epsilon )
+     *   ret[0], ret[1].asWkt(), ret[2]
+     *   # (False, 'Point (0 2)', True)
+     *   ret = QgsGeometryUtils.segmentIntersection( QgsPoint( 0, 0 ), QgsPoint( 0, 5 ), QgsPoint( 0, 2 ), QgsPoint( 1, 5 ), epsilon, True )
+     *   ret[0], ret[1].asWkt(), ret[2]
+     *   # (True, 'Point (0 2)', True)
+     *   ret = QgsGeometryUtils.segmentIntersection( QgsPoint( 0, -5 ), QgsPoint( 0, 5 ), QgsPoint( 2, 0 ), QgsPoint( -1, 0 ), epsilon )
+     *   ret[0], ret[1].asWkt(), ret[2]
+     *   # (True, 'Point (0 0)', True)
+     * \endcode
      */
-    static bool segmentIntersection( const QgsPoint &p1, const QgsPoint &p2, const QgsPoint &q1, const QgsPoint &q2, QgsPoint &inter SIP_OUT, double tolerance );
+    static bool segmentIntersection( const QgsPoint &p1, const QgsPoint &p2, const QgsPoint &q1, const QgsPoint &q2, QgsPoint &intersectionPoint SIP_OUT, bool &isIntersection SIP_OUT, double tolerance, bool acceptImproperIntersection = false );
```

Comment:
may I ask another small change? 
it makes more sense to keep tolerance at the end of list and we could add some default value there, no?

Reply:
Yes to add a default value, but not at the end of the list.
Since tolerance is more used than the newest parameter, I think it's better to keep the other at the end. There is more tools using or that will use this method which define also a tolerance to use, not necessarily with the need to know if there is a proper/improper intersection.